include/pybind11/detail/internals.h

```diff
@@ -168,15 +173,31 @@ struct override_hash {
     }
 };
 
+using instance_map = std::unordered_multimap<const void *, instance *>;
+
+struct instance_map_shard {
+    std::mutex mutex;
+    instance_map registered_instances;
+    char padding[64 - (sizeof(std::mutex) + sizeof(instance_map)) % 64];
```

Comment:
maybe `alignas()` for the struct instead? That also adds padding, and then you get alignment on top.